### ISSUE TITLE
Support for `frequency_penalty`

### ISSUE THREAD

**Issue by richardliaw** [NVIDIA/TensorRT-LLM#275]
>Frequency Penalty is between -2.0 and 2.0 and it impacts how the model penalizes new tokens based on their existing frequency in the text.

> Positive values will decrease the likelihood of the model repeating the same line verbatim by penalizing new tokens that have already been used frequently.

This parameter is important for OpenAI compatibility, which is a growing standard for LLM usage.

cc @Yard1 @akshay-anyscale

**Comment by juney-nvidia** [comment]
@richardliaw 

Thanks for reporting this. Can you elaborate a little bit more about the difference between the already supported [repetiton_penalty](https://nvidia.github.io/TensorRT-LLM/gpt_runtime.html#sampling-parameters) and `frequency_penalty` here? 

Thanks
June

**Comment by richardliaw** [comment]
Frequency penalty has a specific implementation here: https://platform.openai.com/docs/guides/gpt/parameter-details.

vLLM treats the two differently (repetition_penalty is multiplicative with the logits, whereas frequency penalty is additive): https://github.com/vllm-project/vllm/blob/9f669a9a7c2b2d0a7963a6e29253280e57680adb/vllm/model_executor/layers/sampler.py#L233-L236

**Comment by juney-nvidia** [comment]
Thanks for sharing this, this issue relates to [the other](https://github.com/NVIDIA/TensorRT-LLM/issues/274) one, both regarding to the control of sampling/decoder process. We will follow up and reply later.

June

**Comment by npuichigo** [comment]
Same issue here. I'm also facing the issue to delegate `frequency_penalty` to the trtllm backend. 
https://github.com/npuichigo/openai_trtllm/blob/6bba4967fbcce187b14e89043126a878ca872507/src/routes/chat.rs#L277-L280

**Comment by kaiyux** [comment]
The support should be included in the latest main branch, please see https://github.com/NVIDIA/TensorRT-LLM/pull/754.

Closing. Please feel free to comment if you have any question, thanks!
